### PR TITLE
ARROW-9280: [Rust] [Parquet] Calculate page and column statistics

### PR DESCRIPTION
**Pull request by zeevm**
1. Calculate page and column statistics
2. Use pre-calculated statistics when available to speed-up when writing data from other formats like ORC.

### CODE REVIEW

**Comment by nevi-me** [comment]
Hi @zeevm, may you please kindly rebase (to fix the Rust failures) and open a JIRA for this PR

**Comment by zeevm** [comment]
Opened a Jira.

Rebased but the build still fails

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows 10

**Comment by nevi-me** [comment]
Thanks @zeevm, the remaining failure can be fixed by running `cargo +stable fmt` from within the rust folder. We use the `rustfmt` from the stable toolchain, to avoid more frequent changes from a nightly one.

File: rust/parquet/src/column/writer.rs [hunk 1]
```diff
@@ -216,26 +278,26 @@ impl<T: DataType> ColumnWriterImpl<T> {
             def_levels_sink: vec![],
             rep_levels_sink: vec![],
             data_pages: VecDeque::new(),
+            min_page_value: None,
+            max_page_value: None,
+            num_page_nulls: 0,
+            page_distinct_count: None,
```
Comment (sunchao) [inline rust/parquet/src/column/writer.rs:284]: Seems this is not used at all?

File: rust/parquet/src/column/writer.rs [hunk 2]
```diff
@@ -387,15 +538,28 @@ impl<T: DataType> ColumnWriterImpl<T> {
             ));
         }
 
-        // TODO: update page statistics
+        if calculate_page_stats {
+            for val in &values[0..values_to_write] {
+                if self.min_page_value.is_none()
```
Comment (sunchao) [inline rust/parquet/src/column/writer.rs:543]: maybe we should extract this into a util function `update_page_min_max` and similarly for column min/max value.

**Review by sunchao — COMMENTED** [review]
Thanks @zeevm - left some comments.

File: rust/parquet/src/column/writer.rs [hunk 3]
```diff
@@ -276,12 +372,60 @@ impl<T: DataType> ColumnWriterImpl<T> {
...
+
+    /// Writer may optionally provide pre-calculated statistics for this batch, in which case we do
+    /// not calculate page level statistics as this will defeat the purpose of speeding up the write
+    /// process with pre-calculated statistics.
+    pub fn write_batch_with_statistics(
+        &mut self,
+        values: &[T::T],
+        def_levels: Option<&[i16]>,
+        rep_levels: Option<&[i16]>,
+        min: &Option<T::T>,
```
[the remark this answers is not in the document]
Comment (zeevm) [reply rust/parquet/src/column/writer.rs:388]: IIUC the format specifies that the various stats are optional so it seems reasonable to allow the caller to specify only some of the values isn't it?

File: rust/parquet/src/column/writer.rs (hunk 1 above)
Comment (zeevm) [reply rust/parquet/src/column/writer.rs:259]: These are used to track page level stats and write those stats when writing a page and to update the column level stats when writing the page.

Comment (sunchao) [reply rust/parquet/src/column/writer.rs:259]: Yes. But it is not updated nor used at all. Can you double check?

Comment (zeevm) [reply rust/parquet/src/column/writer.rs:259]: Are you specifically referring to 'page_distinct_count', or all 4 page level vars?

Comment (zeevm) [reply rust/parquet/src/column/writer.rs:259]: All of them are used, page_distinct_count isn't being calculated in this PR though, probably a following PR

Comment (zeevm) [reply rust/parquet/src/column/writer.rs:259]: They're used here:

flush_data_pages()
make_typed_statistics()
update_page_min_max()

Comment (sunchao) [reply rust/parquet/src/column/writer.rs:259]: Yes I meant `page_distinct_count`. It is fine to do this in a follow upo.

**Review by sunchao — APPROVED** [review]
LGTM

File: rust/parquet/src/column/writer.rs (hunk 1 above)
Comment (zeevm) [reply rust/parquet/src/column/writer.rs:259]: How do I merge? Do I need to ask for merge/write permissions to complete the PR?

**Comment by zeevm** [comment]
How do I complete/merge the PR, do I need to request write access or will it be completed by a maintainer?

**Comment by sunchao** [comment]
@zeevm once approved, a committer will help merge this. Seems the PR now is a little messed up, can you clean it up so I can merge it?

**Pull request closed without merging**